### PR TITLE
Refactored shadowmapping.

### LINKED ISSUE #262: Improve shadow quality (currently shadow map is underused)
**Describe the project you are working on:**

I am working on a first person shooter for mobile. It involves rendering shadows.

**Describe the problem or limitation you are having in your project:**

I need to keep the shadow map resolution low (say 1024x1024) to run at full speed on mobile.
At this resolution, the quality of the Godot shadows is much lower than what we could expect.

Below the quality with the **Godot shadow rendering pipeline**

godot

Below the quality with a **custom shadow rendering pipeline** (written in GDScript)
custom

**Describe how this feature / enhancement will help you overcome this problem or limitation:**

An improvement of the shadow quality to an acceptable level will allow me not to tinker with the rendering pipeline ;)

**Show a mock up screenshots/video or a flow diagram explaining how your proposal will work:**

**Describe implementation detail for your proposal (in code), if possible:**

Source code of the custom implementation: https://github.com/galmiza/godot-custom-shadow-mapping

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**

**Is there a reason why this should be core and not an add-on in the asset library?:**

This enhancement is expected to replace the current implementation.

### LINKED ISSUE #11398: Subsurface scattering does not support transmittance.
**Operating system or device, Godot version, GPU Model and driver (if graphics related):**
Fedora 26, Nvidia Proprietary 384.69  bc005a57d38137b062e0ccc26467353f53556930

**Issue description:**

SSS does not support transmittance, here is the SSS demo from sketchfab

screenshot from 2017-09-18 14-57-13
does not look really nice, but it lets light through thinner parts of the object,

here is how it looks in godot
screenshot from 2017-09-18 15-04-57
screenshot from 2017-09-18 15-05-43

ping @reduz 

**Steps to reproduce:**

**Link to minimal example project:**

### LINKED ISSUE #12148: Shadows create artifacts in Godot 3.0
**Operating system or device, Godot version, GPU Model and driver (if graphics related):**
Manjaro 17.0.5
Godot 3.0 Master Build
NVidia GT540M, driver 384.90

**Issue description:**

Small cubes from OmniLight shadow.
screenshot_2017-10-16_15-04-11

**Steps to reproduce:**
Just create an Omni Light above a "floor" and anable shadow.

**Link to minimal example project:**

[shadows.zip](https://github.com/godotengine/godot/files/1387324/shadows.zip)

### LINKED ISSUE #13575: PSSM artifacts with directional lights
**Operating system or device, Godot version, GPU Model and driver (if graphics related):**
godot master branch

**Issue description:**

It seems to me pssm and/or pcf need a little more work to be really usable.

For a simple scene with a cube and sphere, the default settings are not great (gap due to bias) but even after many tweaking I would say it's not ok : 
- split incoherence
- jaggy resolution (the settings for resolution in project settings do nothing)

**Steps to reproduce:**
Here my little scene with results (which are very different in every viewports) :
screenshot from 2017-12-01 23-42-06

Here a comparison with unity (with no tweaking) :
screenshot from 2017-12-01 23-58-02

**Link to minimal example project:**

### LINKED ISSUE #17260: Shadow Normal Offset Bias
Normal Offset is a better way to eliminate self-shadowing with much less to virtually no peter panning:
gdc_poster_normaloffset

video explanation: https://web.archive.org/web/20160602213857/https://www.dissidentlogic.com/old/video/Techniques/NormalOffsetShadows_0001.wmv

Note: Godot's "Normal bias" seems to be slope bias, and it introduces a lot of peter panning.

### LINKED ISSUE #23603: Window width affects directional shadow resolution.
**Godot version:**
Affects both 3.0.6 and Official 3.1 Alpha 2

**OS/device including version:**
Linux Mint 19 - 64bit
Nvidia GeForce GTX 1060 on driver version: 396.54

**Issue description:**
For some reason the directional shadow resolution decreases as the window width increases, and increases as the width decreases.
Happens in the editor window and also in a running game.
[Video](https://github.com/godotengine/godot/files/2563615/what.flv.zip)

I'm assuming this isn't intended behaviour?
OmniLight and SpotLight do not seem affected.

**Steps to reproduce:**
1. Set up a scene with meshes and a Directional Light with shadows enabled.
2. Decrease and increase the width of the 3D scene window and observe shadows.

### LINKED ISSUE #24587: Shadow atlas size setting does not change resolution
**Godot version: 3.1 alpha 4, also appears in 3.0.6 stable**

**OS/device including version: Windows 10, GTX 1070ti, NVidia 417.35 Drivers**

**Issue description:**
Changing the shadow atlas size appears to affect the softness of the filtering, rather than increasing the resolution of the shadow.

I've tested the default project quality settings with shadow filtering set to PCF13, using shadow atlas sizes from 512 by 512 to 16384 by 16384. The scene is a cuboid and a plane, with one omni light casting a shadow. It looks like the shadow just gradually loses the effect of the filtering as the resolution increases, rather than increasing the resolution itself.

512 by 512:

1024 by 1024:

2048 by 2048:

4096 by 4096:

8192 by 8192:

16384 by 16384:

I can't tell by eye if the resolution is actually changing as expected at the lower resolutions, but it's clear to me that the shadows are exactly the same between 4096, 8192 and 16384, aside from the higher two looking worse due to not having the edges smoothed as much. Currently, it doesn't seem worth it to turn shadows to anything other than somewhere between 2048 or 1024, as those seem to be the best balance between clarity and smooth edges.

**Steps to reproduce:**
1. Set shadow filtering to PCF13.
1. Set shadow atlas resolution to any low power of 2 in project settings, then raise until hitting the max.
1. Observe how shadow blockiness never changes (aside from being imperceptible at the lower resolutions due to excessive smoothing), but smooth filtering decreases as resolution increases.

**Minimal reproduction project:**
[Shadow Test.zip](https://github.com/godotengine/godot/files/2707801/Shadow.Test.zip)
The only setting I changed between the screenshots was shadow atlas size. It's 
currently set to max in this project. The omni light has its bias set to 0.05, but is otherwise unchanged.

### LINKED ISSUE #29690: Shadows from DirectionalLight become jittery further away.
**Godot version:**
3.1.1

**OS/device including version:**
Ubuntu 18.04
Intel® Core™ i5-6600K CPU @ 3.50GHz × 4
GeForce GTX 1060 6GB/PCIe/SSE2

**Issue description:**
Our scene is rather large and I've noticed that the shadows are quite jittery round the edges. I've narrowed it down to it being caused by the large offset to the zero point. It does not seem to matter how many objects there are in the scene, just that the object is very far removed from the 0,0,0 point. While the in-game camera or editor camera is moving you will see a jittery border around the shadows.

**Steps to reproduce:**
Run BugReport project.
The animation shows the shadows at both close and far away.

**Minimal reproduction project:**
[BugReport.zip](https://github.com/godotengine/godot/files/3278423/BugReport.zip)

### LINKED ISSUE #34010: OmniLight Pixelated Shadows artifact
**Godot version:**
Godot 3.2 Beta

**OS/device including version:**
Windows 7 x64bit

**Issue description:**
OmniLight have a Weird Pixelated Shadows Like this :
even if increasing Shadow Atlas size

editor_screenshot_2019-11-30T015304+0100

and the issue is huge if those cubes are moving

**Minimal reproduction project:**
[TestLightShadow.zip](https://github.com/godotengine/godot/files/3906622/TestLightShadow.zip)

### PR DESCRIPTION
**Pull request by reduz**
image

- Made shadow bias size independent, so it will remain when changing light or camera size.
- Implemented normal offset bias, which greatly enhances quality.
- Added transmission to subsurface scattering

Closes #17260
*Bugsquad edit: closes https://github.com/godotengine/godot-proposals/issues/262, fixes #34010, fixes #29690, fixes #23603, fixes #12148, fixes #13575, fixes #24587, fixes #11398*

### CODE REVIEW

File: scene/3d/light_3d.cpp [hunk 1]
```diff
@@ -414,11 +416,8 @@ void DirectionalLight3D::_bind_methods() {
 DirectionalLight3D::DirectionalLight3D() :
 		Light3D(RenderingServer::LIGHT_DIRECTIONAL) {
 
-	set_param(PARAM_SHADOW_NORMAL_BIAS, 0.8);
```
Comment (clayjohn) [inline scene/3d/light_3d.cpp:417]: You removed the defaults here and then never replaced them

**Comment by Zireael07** [comment]
That's closing A LOT of issues :wink:

File: servers/rendering/rasterizer_rd/shaders/scene_high_end.glsl [hunk 2]
```diff
@@ -267,6 +261,14 @@ VERTEX_SHADER_CODE
 #else
 	gl_Position = projection_matrix * vec4(vertex_interp, 1.0);
 #endif
+
+#ifdef MODE_RENDER_DEPTH
+	if (scene_data.pancake_shadows) {
+		if (gl_Position.z <= 0.00001) {
```
Comment (clayjohn) [inline servers/rendering/rasterizer_rd/shaders/scene_high_end.glsl:267]: Shouldn't this be a variable amount based on the pancake size?

File: servers/rendering/rasterizer_rd/shaders/scene_high_end.glsl [hunk 3]
```diff
@@ -728,24 +792,81 @@ void light_process_omni(uint idx, vec3 vertex, vec3 eye_vec, vec3 normal, vec3 a
...
 		} else {
-
 			splane.z = 1.0 - splane.z;
 		}
 
 		splane.xy /= splane.z;
+
+		{
+
+#if 0
```
Comment (clayjohn) [inline servers/rendering/rasterizer_rd/shaders/scene_high_end.glsl:802]: Is this intentionally left out?

File: scene/3d/light_3d.cpp (hunk 1 above)
Comment (reduz) [reply scene/3d/light_3d.cpp:417]: ah..

File: servers/rendering/rasterizer_rd/shaders/scene_high_end.glsl (hunk 2 above)
Comment (reduz) [reply servers/rendering/rasterizer_rd/shaders/scene_high_end.glsl:267]: no.. the pancake is applied to the orthogonal frustum basically by making it X distance larger towards the light from the slice sphere, so all you have to do is clamp all geometry to the beginning.

File: servers/rendering/rasterizer_rd/shaders/scene_high_end.glsl (hunk 3 above)
Comment (reduz) [reply servers/rendering/rasterizer_rd/shaders/scene_high_end.glsl:802]: ah yes, this is from a previous attempt to implement normal bias in this type of light which I wasn' t very convinced about. The one above works better.

**Comment by fire** [comment]
How difficult is the backport of the code for 3.2?

**Comment by Calinou** [comment]
> Is there a patchset or fork that brings these changes to 3.x? There's probably no way this would be in an official 3.x release since the changes to shadow/light properties would break compatibility, but it would be great if I could take advantage of the improved shadows now instead of Waiting For Godot 4.1. The shadows in 3.x are serviceable but could be a lot sharper and there isn't a way to avoid both artifacting and peterpanning. The shadows in 4 on the other hand are basically magic.

Se https://github.com/godotengine/godot/pull/43207.

**Comment by cosmicchipsocket** [comment]
Shadows in 4.0 are definitely a bit sharper with the same settings versus 3.x (and a _lot_ sharper with a wider aspect ratio). But the big thing for me is that with 3.x I can get simultaneous artifacting and peterpanning, whereas with 4.0 I can avoid both.

Shadows_PCF5_3_x
Shadows_PCF5_4_0

This is with 4k textures in both.

I know about the reverse culling option, but it adds artifacting to sharp corners, and for my visual style I don't want to subdivide everything. I want the graphics looking sharp and clean.

**Pull request merged**